devito/passes/iet/mpi.py

```diff
@@ -54,123 +54,176 @@ def _drop_reduction_halospots(iet):
     return iet
 
 
-def _hoist_invariant(iet):
+def _hoist_redundant_from_conditionals(iet):
     """
-    Hoist HaloSpots from inner to outer Iterations where all data dependencies
-    would be honored. This pass avoids redundant halo exchanges when the same
-    data is redundantly exchanged within the same Iteration tree level.
-
-    Example:
-                                   haloupd v[t0]
-    for time                       for time
-      W v[t1]- R v[t0]               W v[t1]- R v[t0]
-      haloupd v[t1]                  haloupd v[t1]
-      R v[t1]                        R v[t1]
-      haloupd v[t0]                  R v[t0]
-      R v[t0]
-
+    Hoist redundant HaloSpots from Conditionals. The idea is that, by doing so,
+    the subsequent passes (hoist, merge) will be able to optimize them away.
+
+    Examples
+    --------
+
+    for time                          for time
+      if(cond)                          if(cond)
+        haloupd u[t0]                     haloupd u[t0]
+        haloupd v[t0]                   haloupd v[t0]
+      W v[t1]- R u[t0],v[t0]   --->     W v[t1]- R u[t0],v[t0]
+      haloupd v[t0]                     haloupd v[t0]
+      R v[t0]                           R v[t0]
+
+    Note that in the above example, the `haloupd v[t0]` in the if branch is
+    redundant, as it also appears later on in the Iteration body, so it is
+    hoisted out of the Conditional.
     """
-
-    # Precompute scopes to save time
-    scopes = {i: Scope([e.expr for e in v]) for i, v in MapNodes().visit(iet).items()}
-
-    # Analysis
-    hsmapper = {}
-    imapper = defaultdict(list)
-
     cond_mapper = _make_cond_mapper(iet)
     iter_mapper = _filter_iter_mapper(iet)
 
+    mapper = HaloSpotMapper()
     for it, halo_spots in iter_mapper.items():
-        for hs0, hs1 in combinations(halo_spots, r=2):
-            if _check_control_flow(hs0, hs1, cond_mapper):
-                continue
+        scope = Scope([e.expr for e in FindNodes(Expression).visit(it)])
 
-            # If there are overlapping loc_indices, skip
-            hs0_mdims = hs0.halo_scheme.loc_values
-            hs1_mdims = hs1.halo_scheme.loc_values
-            if hs0_mdims.intersection(hs1_mdims):
+        for hs0 in halo_spots:
+            conditions = cond_mapper[hs0]
+            if not conditions:
                 continue
+            condition = conditions[-1]  # Take the innermost Conditional
 
-            for f, v in hs1.fmapper.items():
-                if f not in hs0.functions:
-                    continue
+            for f in hs0.fmapper:
+                hsf0 = hs0.halo_scheme.project(f)
+
+                # Find candidate for subsequent merging
+                for hs1 in halo_spots:
+                    if hs0 is hs1 or cond_mapper[hs1]:
```

Comment:
Could check if `cond_mapper[hs1]` is a "subset" of "cond_mapper[hs0]" , i.e less conditions in which case it could be hoisted

Reply:
I could, but that really is an impossible situation